toolchain/check/member_access.cpp

```diff
@@ -98,6 +102,79 @@ static auto IsInstanceMethod(const SemIR::File& sem_ir,
   return false;
 }
 
+// Returns the FunctionId of the current function if it exists.
+static auto GetCurrentFunction(Context& context)
+    -> std::optional<SemIR::FunctionId> {
+  if (context.return_scope_stack().empty()) {
+    return std::nullopt;
+  }
+
+  return context.insts()
+      .GetAs<SemIR::FunctionDecl>(context.return_scope_stack().back().decl_id)
+      .function_id;
+}
+
+static auto GetHighestAllowedAccess(Context& context, Parse::NodeId node_id,
+                                    SemIR::ConstantId name_scope_const_id)
+    -> SemIR::AccessKind {
+  auto current_function = GetCurrentFunction(context);
```

Comment:
```suggestion
  // TODO: Maybe use LookupUnqualifiedName for `Self` to support things like
  // `var x: Self.ParentProtectedType`?
  auto current_function = GetCurrentFunction(context);
```

Just a note. I think it's fine to only support functions for now.